src/connectionpools.jl

```diff
@@ -94,128 +95,64 @@ This new connection instance will be tracked by the `Pod` and MUST be returned t
+            conn !== nothing && close(conn)
+            conn = f()
         end
-    finally
-        unlock(pod.lock)
     end
+    return conn
 end
 
+taskid() = string(objectid(current_task()) % UInt16, base=16, pad=4)
+
 function release(pod::Pod{C}, conn::C; return_for_reuse::Bool=true) where {C}
-    lock(pod.lock)
+    if return_for_reuse && isopen(conn)
+        push!(pod.connections, conn)
+    else
+        close(conn)
+    end
+    lock(pod.cond) # acquire barrier
     try
-        # We first want to look up the corresponding ConnectionTracker object in our
-        # Pod `active` Dict that tracks active connections
-        id = connectionid(conn)
-        # if, for some reason, it's not in our `active` tracking Dict
-        # then something is wrong; you're trying to release a `Connection`
-        # that this Pod currently doesn't think is active
-        if !haskey(pod.active, id)
-            error("couldn't find connection id in pod's current list of active connections; invalid release call; each acquired connection should be `release`ed ONLY once")
-        end
-        conn_tracker = pod.active[id]
-        # remove the ConnectionTracker from our `active` Dict tracker
-        delete!(pod.active, id)
-        if return_for_reuse && isopen(conn)
-            # reset the idle_timestamp of the ConnectionTracker
-            conn_tracker.idle_timestamp = time()
-            # check if there are any tasks waiting to acquire a connection
-            if isempty(pod.lock)
-                # if not, we put the connection back in the pod idle queue
-                # in this case, there's no need to notify the pod lock/condition
-                # since there's no one waiting to be notified anyway
-                # println("$(taskid()): returning connection (id='$(_id(id))') to pod idle_timeout queue for reuse")
-                push!(pod.conns, conn_tracker)
-            else
-                # if there are waiters, we notify the pod condition and pass the
-                # ConnectionTracker object in the notification; we ensure to pass
-                # all=false, so only one waiter is woken up and receives the
-                # ConnectionTracker
-                # println("$(taskid()): returning connection (id='$(_id(id))') to a waiting task for reuse")
-                notify(pod.lock, conn_tracker; all=false)
-            end
-        else
-            # if the user has, for whatever reason, requested this connection not be reused
-            # anymore by passing `return_for_reuse=false`, then we've still removed it from
-            # the `active` tracking and want to notify the pod in case there are waiting
-            # acquire tasks that can now create a new connection
-            # println("$(taskid()): connection not reuseable; notifying pod that a connection has been released though")
-            notify(pod.lock, nothing; all=false)
+        if notify(pod.cond; all=false) == 0
```

Comment:
Let's maybe add a comment about how the fastpath works here with the waiters in `acquire`. It makes sense once you get it, but it's not easy to understand without guidance. Also the `pod.cond` could use some comments outlining how it is used.

Reply:
good call